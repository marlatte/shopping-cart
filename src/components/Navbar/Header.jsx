import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import CartButton from './CartButton';
import NavMenu from './NavMenu';
import HomeLink from '../ui/HomeLink';

export default function Header({ items }) {
  const [menuOpen, setMenuOpen] = useState(false);

  useEffect(() => {
    const focusedClassName = menuOpen ? '.close-menu-btn' : '.open-menu-btn';
    document.querySelector(focusedClassName).focus();
  }, [menuOpen]);

  return (
    <header>
      {menuOpen && (
        <NavMenu
          onClick={() => {
            setMenuOpen(false);
          }}
        />
      )}
      <button
        type="button"
        aria-label="Menu"
        className="open-menu-btn"
        onClick={() => {
          setMenuOpen(true);
        }}
      >
        â˜°
      </button>
      <HomeLink />
      <CartButton items={items} />
    </header>
  );
}

Header.propTypes = {
  items: PropTypes.number.isRequired,
};
