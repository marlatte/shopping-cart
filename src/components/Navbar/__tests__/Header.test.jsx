import { expect, test } from 'vitest';
import { BrowserRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Header from '../Header';

function setup(items = 0) {
  return render(<Header items={items} />, {
    wrapper: BrowserRouter,
  });
}

test('has 3 children by default: a title, mini-cart, and menu button', () => {
  setup();
  expect(screen.getByRole('banner').childElementCount).toBe(3);
});

test('logo is link to homepage', () => {
  setup();
  expect(
    screen.getByRole('link', { name: 'Yuedpao Homepage' })
  ).toHaveAccessibleName(/yuedpao homepage/i);
});

test('menu button shows 3 bars but is named "menu"', () => {
  setup();
  expect(screen.getByRole('button')).toHaveAccessibleName(/menu/i);
});

test('Matches snapshot for title, mini-cart, and menu button', () => {
  const { container } = setup();
  expect(container).toMatchSnapshot();
});

test('clicking the menu button opens the NavMenu', async () => {
  const user = userEvent.setup();
  setup();
  const menuBtn = screen.getByRole('button', { name: 'Menu' });

  await user.click(menuBtn);

  expect(screen.getByRole('banner').childElementCount).toBe(5);
});

test('clicking the close button closes the NavMenu', async () => {
  const user = userEvent.setup();
  setup();

  const menuBtn = screen.getByRole('button', { name: 'Menu' });
  await user.click(menuBtn);

  const closeBtn = screen.getByRole('button', { name: 'Close Menu' });
  await user.click(closeBtn);

  expect(screen.getByRole('banner').childElementCount).toBe(3);
});
