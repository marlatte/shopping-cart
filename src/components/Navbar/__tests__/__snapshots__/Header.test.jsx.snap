// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Matches snapshot for title, mini-cart, and menu button 1`] = `
<div>
  <header>
    <button
      aria-label="Menu"
      class="open-menu-btn"
      type="button"
    >
      â˜°
    </button>
    <a
      href="/"
    >
      <span
        aria-label="Yuedpao Homepage"
        class="logo"
        title="Yuedpao Homepage"
      >
        Yuedpao
      </span>
    </a>
    <a
      href="/cart"
    >
      Cart
    </a>
  </header>
</div>
`;
